tiny_dnn/layers/cell.h

```diff
@@ -31,16 +31,16 @@ class cell: public layer {
                         std::vector<tensor_t *> &out_grad,
                         std::vector<tensor_t *> &in_grad) = 0;
 
-  virtual core::backend_t get_backend_type() const { return backend_type_; };
+  virtual core::backend_t backend_type() const { return wrapper_->engine(); };
 
-  virtual void init_backend(Device * device) = 0;
-
-  inline void set_wrapper(layer* layer_p) { layer_ = layer_p; }
+  virtual void init_backend(const layer* wrapper) = 0;
 
 
 protected:
-  core::backend_t backend_type_;
-  layer * layer_; // every forward iteration, we must get the engine, backend, etc from the wrapper
+
+  inline void set_wrapper(const layer* wrapper) { wrapper_ = wrapper; }
```

Comment:
I tried but since layer* comes from a ``this`` it involves a ``shared_from_this`` and it overcomplicates things